clog/logger.go

```diff
@@ -0,0 +1,238 @@
+package clog
+
+import (
+	"context"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/alcionai/clues"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// Yes, we just hijack zap for our logging needs here.
+// This package isn't about writing a logger, it's about
+// adding an opinionated shell around the zap logger.
+var (
+	cloggerton *clogger
+	singleMu   sync.Mutex
+)
+
+type clogger struct {
```

Comment:
I think we need to be careful about hiding all logger details from callers. When we switched to using clog, we actually broke some connections others had made that allowed us to get logs from another go package we were using. Those connections were made by passing a function that returned a sugared logger to the other go package. Once we switched to clog, we no longer had a way to get a sugered logger, and thus couldn't get logs from the other system when we wanted to

Reply:
Huh, I don't recall that event.  Certainly not an intended result, and I would have been happy to expose the support for that as necessary.  Was this with kopia?  

My long term preference here is to allow the user to retrieve the underlying logger as either an interface (`GetLogger() CommonLoggingInterface`) or by its source struct (if necessary, though retrieving from an interface should be sufficient).  There's also some future design work that needs to happen for managing log message multiplexing (eg: logging to both Zap and OTEL) which will invariably change the construction, storage, and retrieval of the underlying logger.

Reply:
> Was this with kopia?

yes, changes to how logging was initialized broke support for outputting kopia logs but confusingly left the flag to control the old behavior around

> Certainly not an intended result

In the broader scheme of things, it's not really a question of whether it was intended or not, as the assumption myself (and I would think others) are operating under is that changes are well intentioned. The unfortunate thing is that, to borrow a phrase I saw in a book, good intentions don't guarantee good outcomes and that breaking changes like that are disruptive to the team since now people spend time debugging and writing/waiting for fixes instead of the thing they originally wanted to do

> My long term preference here is to allow the user to retrieve the underlying logger as either an interface (GetLogger() CommonLoggingInterface) or by its source struct

I'm not sure if I have a clear enough picture on other systems that this may need to interact with yet to have input here. The bottom line though is that there will need to be some way to allow getting logs from other systems to where we need them